crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+    let dim = umat.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+    let half_dim = dim / 2;
+    let umat_array: Array2<Complex64> = Array2::from_shape_fn(umat.shape(), |(i, j)| umat[(i, j)]);
+    let mut utensor = umat_array
+        .to_shape(vec![2; 2 * num_qubits])
+        .unwrap()
+        .into_owned();
+    if k != 0 {
+        utensor = move_axis(utensor, k, 0);
+        utensor = move_axis(utensor, k + num_qubits, num_qubits);
+    }
+    let ud4 = utensor.to_shape([2, half_dim, 2, half_dim]).unwrap();
+    let um00 = ud4.slice(s![0, .., 0, ..]);
+    let um11 = ud4.slice(s![1, .., 1, ..]);
+    let um01 = ud4.slice(s![0, .., 1, ..]);
+    let um10 = ud4.slice(s![1, .., 0, ..]);
+    [
+        DMatrix::from_fn(um00.shape()[0], um00.shape()[1], |i, j| um00[[i, j]]),
+        DMatrix::from_fn(um11.shape()[0], um11.shape()[1], |i, j| um11[[i, j]]),
+        DMatrix::from_fn(um01.shape()[0], um01.shape()[1], |i, j| um01[[i, j]]),
+        DMatrix::from_fn(um10.shape()[0], um10.shape()[1], |i, j| um01[[i, j]]),
+    ]
+}
+
+/// Checks whether off-diagonal blocks are zero.
+fn off_diagonals_are_zero(
+    um01: &DMatrix<Complex64>,
+    um10: &DMatrix<Complex64>,
+    atol: Option<f64>,
+) -> bool {
+    let atol = atol.unwrap_or(1e-12);
+    um01.iter()
+        .all(|x| abs_diff_eq!(*x, Complex64::ZERO, epsilon = atol))
+        && um10
+            .iter()
+            .all(|x| abs_diff_eq!(*x, Complex64::ZERO, epsilon = atol))
+}
+
+/// The optimization A.2 from [1, 2]. This decomposes two qubit unitaries into a
+/// diagonal gate and a two cx unitary and reduces overall ``cx`` count by
+/// 4^(n-2) - 1. This optimization should not be done if the original unitary is controlled.
+fn apply_a2(
+    circ: &CircuitData,
+    two_qubit_decomposer: &TwoQubitBasisDecomposer,
+) -> PyResult<CircuitData> {
+    let ind2q: Vec<usize> = circ
+        .data()
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, inst)| {
+            if matches!(inst.op.view(), OperationRef::Unitary(_)) {
+                if let Some(ref label) = inst.label {
+                    if label.as_str() == "qsd2q" {
+                        return Some(idx);
+                    }
+                }
+            }
+            None
+        })
+        .collect();
+    if ind2q.is_empty() {
+        return Ok(circ.clone());
+    } else if ind2q.len() == 1 {
+        let mut circ = circ.clone();
+        circ.invalidate_label(ind2q[0]);
+    }
+    let mut diagonal_rollover: HashMap<usize, CircuitData> = HashMap::with_capacity(ind2q.len());
+    let mut new_matrices: HashMap<usize, Array2<Complex64>> = ind2q
+        .iter()
+        .map(|idx| {
+            let OperationRef::Unitary(unitary) = circ.data()[*idx].op.view() else {
+                unreachable!("diagonal unitary is not a unitary gate");
+            };
+            (*idx, unitary.matrix_view().to_owned())
```

Comment:
You can just do:

```suggestion
            (*idx, unitary.matrix().unwrap())
```

`.matrix_view()` is to get a no-copy read-only ndarray view of the unitary, while we are explicitly copying here. We don't actually need owned matrices for this outer initialization, but I don't think the matrices generated by `two_qubit_decompose_up_to_diagonal` will live long enough to use a view, so it's fine to use owned here I guess.

Reply:
Note that `matrix()` accepts `_params: &[Param]`; I passed `&[]`, hope it's ok.